From d248f4265c1c5bc2e30ab03a8ede1f34c1eebc7a Mon Sep 17 00:00:00 2001
From: Naveen Mamindlapalli <naveenm@marvell.com>
Date: Wed, 19 Jan 2022 16:45:47 +0530
Subject: [PATCH 09/12] octeontx2-bphy-netdev: rfoe: fix ptp4l bad message
 issue

commit 135a64cdd4e60c6f8326f0afb3831d3060a29ec2 from
git@git.assembla.com:cavium/WindRiver.linux.git

The RFOE RX DMA updates the DMA block size in multiples of 4 bytes.
This creates an issue in case of L2 PTP packets where the ptp4l
application is considering the trailing padded bytes as extra TLV
and throws bad message error. This patch calculates the skb length
from PTP header messageLength field.

Change-Id: I81c16303cacae6506fecdecbf25b8b206d911c5d
Signed-off-by: Naveen Mamindlapalli <naveenm@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/69484
Reviewed-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
Tested-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/bphy/otx2_rfoe.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_rfoe.c b/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_rfoe.c
index 70e03aba8724..f4cf057f398c 100644
--- a/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_rfoe.c
+++ b/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_rfoe.c
@@ -452,6 +452,7 @@ static void otx2_rfoe_process_rx_pkt(struct otx2_rfoe_ndev_priv *priv,
 	int found = 0, idx, len, pkt_type;
 	struct otx2_rfoe_ndev_priv *priv2;
 	struct otx2_rfoe_drv_ctx *drv_ctx;
+	unsigned int ptp_message_len = 0;
 	struct rfoe_psw0_s *psw0 = NULL;
 	struct rfoe_psw1_s *psw1 = NULL;
 	struct net_device *netdev;
@@ -593,6 +594,12 @@ static void otx2_rfoe_process_rx_pkt(struct otx2_rfoe_ndev_priv *priv,
 	skb_put(skb, len);
 	skb->protocol = eth_type_trans(skb, netdev);
 
+	/* remove trailing padding for ptp packets */
+	if (skb->protocol == htons(ETH_P_1588)) {
+		ptp_message_len = skb->data[2] << 8 | skb->data[3];
+		skb_trim(skb, ptp_message_len);
+	}
+
 	if (priv2->rx_hw_tstamp_en) {
 		otx2_rfoe_calc_ptp_ts(priv, &tstamp);
 		skb_hwtstamps(skb)->hwtstamp = ns_to_ktime(tstamp);
-- 
2.34.1

